Removed duplicate ReadResources call

echalone opened the pull request:
The ReadResources method is already called in the ResourceSet constructor (see https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Resources/ResourceSet.cs ) and would be called a second time in the ResXResourceSet constructor.
This leads to a System.ArgumentException "Item has already been added. Key in dictionary: '\<keyname\>'  Key being added: '\<keyname\>'" because the code would try to add every data element of a resx file a second time (first time in the ResourceSet constructor, second time in the ResXResourceSet constructor).
Therefore, this second call needs to be removed from all ResXResourceSet constructors that already call the base constructor (which are both).
You can recreate this issue just by calling
```
new System.Resources.ResXResourceSet("<PathToResx>")
```

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/3575)

RussKie commented:
> This leads to a System.ArgumentException "Item has already been added. Key in dictionary: '' Key being added: ''" because the code would try to add every data element of a resx file a second time (first time in the ResourceSet constructor, second time in the ResXResourceSet constructor).

Thank you. Can you please provide stack traces and steps to reproduce the underlying error we're trying to fix here?

echalone commented:
I acutally provided an example in the comment, but I'll do it again here ;)
So all you have to do, is call the ResXResourceSet constructor trying to load a resx file. I have added two example resx files (as txt files since resx ain't supported) with which one can try this out.
You can either call in a .NET Core 3.1 C# Windows Forms Project
```
new System.Resources.ResXResourceSet("<FilepathToResx>")
```
or simply in PowerShell Core 7:
```
Add-Type -AssemblyName System.Windows.Forms 
New-Object -TypeName "System.Resources.ResXResourceSet" -ArgumentList "<FilepathToResx>"
```
both will fail with .NET Core 3.1 but will not fail in the old .NET Framework 4.5 etc. (The powershell code will work with PowerShell 5 for example)
[Test_SWF4.resx.txt](https://github.com/dotnet/winforms/files/4918794/Test_SWF4.resx.txt)
[Test_SWF2.resx.txt](https://github.com/dotnet/winforms/files/4918795/Test_SWF2.resx.txt)

RussKie commented:
Thank you, it look sensible. Do you think you could add a unit test to verify we don't regress?

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/winforms?pullRequest=3575) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/winforms?pullRequest=3575) before we can accept your contribution.<br/><br/>:x: echalone [sign now](https://cla.dotnetfoundation.org/dotnet/winforms?pullRequest=3575)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/winforms?pullRequest=3575) it.</sub>

echalone commented:
Sorry for the late response, but I had serious tooth problems for the last 3 weeks ;)
I have added unit tests for this change and I'm covering the RESX file loading, retrieving of resource-reader/writer as well as retrieving strings and icon/bitmap/text files, both case sensitive and case insensitive. I'm referencing the existing (icon/bitmap/text) files in "\src\System.Windows.Forms\tests\UnitTests\TestResources\Files" in the RESX files.

For this I've added four RESX files to the root directory of the System.Windows.Forms.Tests project with option "always copy" during build. I can't put them in any sub-directory because during development files referenced in these RESX files will be loaded relative to the RESX file location, and during test run files referenced in these RESX files will be loaded relative to the build output. So the only location working for both scenarios is if those RESX files are in the project root directory.

The four RESX files have different RESX versions (1.3 and 2.0) and different versions of assemblies referenced in the respective RESX file. The four versions of RESX files are for .NET Framework 1 (SWF 1), .NET Framework 2/3 (SWF 2), .NET Framework 4 (SWF 4) and current development version (SWF 42).

I needed to add a phony commit (a comment which was removed again because I used tabulators instead of spaces at the beginning of the line) because the first build had some random failed Unit Test not associated with this PR.
That originally failed build was: https://dev.azure.com/dnceng/public/_build/results?buildId=763566&view=results
and the failed Unit Test was:
```
System.Windows.Forms.Tests.Text.FontMetrics.Font_AdjustForVerticalAlignment(family: \"Times New Roman\", size: 11, bounds: {X=1,Y=1,Width=100,Height=100}, dt: 8, expected: {X=1,Y=84,Width=100,Height=100})
```
After the rebuild due to my additional commit, the build succeeded and all checks have passed now.

The 8 new Unit Tests I added are all in the new file "\src\System.Windows.Forms\tests\UnitTests\System\Resources\ResXResourceSetTests.cs" and are the following:
```
System.Resources.Tests.ResXResourceSetTests.ResXResourceSet_FileConstructor_SWF1_0
System.Resources.Tests.ResXResourceSetTests.ResXResourceSet_FileConstructor_SWF2_0
System.Resources.Tests.ResXResourceSetTests.ResXResourceSet_FileConstructor_SWF4_0
System.Resources.Tests.ResXResourceSetTests.ResXResourceSet_FileConstructor_SWF42_42
System.Resources.Tests.ResXResourceSetTests.ResXResourceSet_StreamConstructor_SWF1_0
System.Resources.Tests.ResXResourceSetTests.ResXResourceSet_StreamConstructor_SWF2_0
System.Resources.Tests.ResXResourceSetTests.ResXResourceSet_StreamConstructor_SWF4_0
System.Resources.Tests.ResXResourceSetTests.ResXResourceSet_StreamConstructor_SWF42_42
```

Hopefully we can merge this PR successfully now :)

RussKie commented:
> I needed to add a phony commit (a comment which was removed again because I used tabulators instead of spaces at the beginning of the line) because the first build had some random failed Unit Test not associated with this PR.
> That originally failed build was: [dev.azure.com/dnceng/public/_build/results?buildId=763566&view=results](https://dev.azure.com/dnceng/public/_build/results?buildId=763566&view=results)
> and the failed Unit Test was:
> 
> ```
> System.Windows.Forms.Tests.Text.FontMetrics.Font_AdjustForVerticalAlignment(family: \"Times New Roman\", size: 11, bounds: {X=1,Y=1,Width=100,Height=100}, dt: 8, expected: {X=1,Y=84,Width=100,Height=100})
> ```
> 
> After the rebuild due to my additional commit, the build succeeded and all checks have passed now.

We have some test instability that we're investigating, and typically a rebuild helps to get over it.

RussKie commented on src/System.Windows.Forms/tests/UnitTests/System/Resources/ResXResourceSetTests.cs:
```diff
@@ -0,0 +1,444 @@
+            }
+        }
+
+        private void ResXResourceSet_TestResX(System.Resources.ResXResourceSet resxFile)
```
A unit test should ideally test a single code path in a single unit. A meta-test introduces a significant maintenance overhead, and is quite undesirable because it makes it harder to see and debug an issue, if/when code breaks.

Let's break it down to a number of small "single responsibility" tests. Something like:

```cs
        [Theory]
        [InlineData("Test_SWF1_0.resx", "SomeMetadataTest1", "Some text for MetaData 1 node")]
        [InlineData("Test_SWF2_0.resx", "Error", @"TestResources\Files\Error.ico")]
        public void ResXResourceSet_CtorFileName_CanReadResources(string resxFileName, string resourceName, string expected)
        {
            ResXResourceSet resxFile = new ResXResourceSet(resxFileName);

            string value = resxFile.GetString(resourceName, true);
            Assert.Equal(expected, value);
        }

        [Theory]
        [InlineData("Test_SWF1_0.resx", "SomeMetadataTest1", "Some text for MetaData 1 node")]
        [InlineData("Test_SWF2_0.resx", "Error", @"TestResources\Files\Error.ico")]
        public void ResXResourceSet_CtorStream(string resxFileName, string resourceName, string expected)
        {
            using FileStream fs = new FileStream(resxFileName, FileMode.Open));
            ResXResourceSet resxFile = new ResXResourceSet(fs);

            string value = resxFile.GetString(resourceName, true);
            Assert.Equal(expected, value);
        }
```

I wouldn't test all possible permutations of parameters - we're only interested in testing that `ResXResourceSet` can still read resources without the duplicated call to `ReadResources()` (that you discovered).
Testing behaviours of `ResourceSet` class is well outside the scope - we don't own this type, and we take it for granted (it belongs to .NET Runtime https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Resources/ResourceSet.cs).

RussKie replied:
(A late thought) I'm not sure if we need 4 different resx files too, I think we could use just one.

echalone replied:
UnitTests now only load resx file and then try to access one specific string resource per Unit Test (name and expected value are defined via Inline Data).
If the expected value is "null" this means that this resource is expected to be missing.
Also, only one RESX file remains (now called TestResources.resx ), using resx version 2.0 and referencing assemblies of framework 4 or above.

echalone commented on src/System.Windows.Forms/tests/UnitTests/System/Resources/ResXResourceSetTests.cs:
```diff
@@ -0,0 +1,444 @@
+    // NB: doesn't require thread affinity
+    public class ResXResourceSetTests : IClassFixture<ThreadExceptionFixture>
+    {
+        #region Resource names and contents
```
Regions have been removed

echalone commented on src/System.Windows.Forms/tests/UnitTests/System/Resources/ResXResourceSetTests.cs:
```diff
@@ -0,0 +1,444 @@
+        public void ResXResourceSet_FileConstructor_SWF1_0()
+        {
+            ResXResourceSet_TestFile("Test_SWF1_0.resx");
+        }
```
Unit tests have been changed from Fact to Theory with InlineData

echalone commented on src/System.Windows.Forms/tests/UnitTests/System/Resources/ResXResourceSetTests.cs:
```diff
@@ -0,0 +1,444 @@
+        {
+            Assert.True(System.IO.File.Exists(fileName), $@"RESX file ""{fileName}"" not found, make sure it's in the root folder of the unit test project");
+
+            System.Resources.ResXResourceSet resxFile = new System.Resources.ResXResourceSet(fileName);
```
Namespaces have been simplified where possible

echalone commented on src/System.Windows.Forms/tests/UnitTests/System/Resources/ResXResourceSetTests.cs:
```diff
@@ -0,0 +1,444 @@
+
+        private void ResXResourceSet_TestFile(string fileName)
+        {
+            Assert.True(System.IO.File.Exists(fileName), $@"RESX file ""{fileName}"" not found, make sure it's in the root folder of the unit test project");
```
As with previous comment: Namespaces have been simplified where possible

echalone commented:
UnitTests were adapted according to your suggestions:
* Removed regions
* Using Theory tests with InlineData instead of Fact tests
* Simplified namespaces where possible
* Simplified UnitTests (removed checks for case sensitive/insensitive, removed checks for GetObject, removed checks for icon/bitmap file)
* Removed 3 of 4 resx files, renamed remaining resx file, only checking with resx file Version 2.0 and referenced assemblies for framework 4 and above

=> This is now resulting in 14 (instead of 8) new unit tests which are significantly smaller and easier to maintain than the previous 8 unit tests

Let me know if there's anything else :)

RussKie commented on src/System.Windows.Forms/tests/UnitTests/System/Resources/ResXResourceSetTests.cs:
```diff
@@ -0,0 +1,77 @@
+
+                string strResXValue = resxSet.GetString(resourceName);
+
+                if (expected == null)
+                {
+                    Assert.Null(strResXValue);
+                }
+                else
+                {
+                    Assert.NotNull(strResXValue);
+                    Assert.Equal(expected, strResXValue);
+                }
```
We can simplify it further:
```suggestion
                Assert.Equal(expected, strResXValue);
```

If `expected is null` the above assertion is the same as `Assert.Null(strResXValue)`; otherwise it assert both that `strResXValue is not null` and `strResXValue == expected`.

In general a unit test should have a cyclomatic complexity of 1 (i.e. no conditional branching) - otherwise you need to write a unit test for a unit test 🤣

RussKie commented on src/System.Windows.Forms/tests/UnitTests/System/Resources/ResXResourceSetTests.cs:
```diff
@@ -0,0 +1,77 @@
+
+                    string strResXValue = resxSet.GetString(resourceName);
+
+                    if (expected == null)
+                    {
+                        Assert.Null(strResXValue);
+                    }
+                    else
+                    {
+                        Assert.NotNull(strResXValue);
+                        Assert.Equal(expected, strResXValue);
+                    }
```
```suggestion
                    Assert.Equal(expected, strResXValue);
```

echalone commented:
Alright, both suggestions are implemented and all Unit Tests remain green, looking good :)
Thanks so far for your support during this pull request ;)

RussKie commented:
Thank you!
